native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/NativeObjCRefinementChecker.kt

```diff
@@ -0,0 +1,159 @@
+        if (objCAnnotations.isNotEmpty() && swiftAnnotations.isNotEmpty()) {
+            swiftAnnotations.forEach {
+                val reportLocation = DescriptorToSourceUtils.getSourceFromAnnotation(it) ?: declaration
+                context.trace.report(ErrorsNative.REDUNDANT_SWIFT_REFINEMENT.on(reportLocation))
+            }
+        }
+        checkOverrides(declaration, descriptor, context, objCAnnotations, swiftAnnotations)
+    }
+
+    private fun DeclarationDescriptor.findRefinedAnnotations(): Pair<List<AnnotationDescriptor>, List<AnnotationDescriptor>> {
+        val objCAnnotations = mutableListOf<AnnotationDescriptor>()
+        val swiftAnnotations = mutableListOf<AnnotationDescriptor>()
+        for (annotation in annotations) {
+            val annotations = annotation.annotationClass?.annotations ?: continue
+            for (metaAnnotation in annotations) {
+                when (metaAnnotation.fqName) {
+                    refinesForObjCFqName -> {
+                        objCAnnotations.add(annotation)
+                        break
+                    }
+                    refinesInSwiftFqName -> {
+                        swiftAnnotations.add(annotation)
+                        break
+                    }
+                }
+            }
+        }
+        return objCAnnotations to swiftAnnotations
+    }
+
+    private fun checkOverrides(
+        declaration: KtDeclaration,
+        descriptor: CallableMemberDescriptor,
+        context: DeclarationCheckerContext,
+        objCAnnotations: List<AnnotationDescriptor>,
+        swiftAnnotations: List<AnnotationDescriptor>
+    ) {
+        if (descriptor.overriddenDescriptors.isEmpty()) return
+        var isRefinedForObjC = objCAnnotations.isNotEmpty()
+        var isRefinedInSwift = swiftAnnotations.isNotEmpty()
+        val supersNotRefinedForObjC = mutableListOf<CallableMemberDescriptor>()
+        val supersNotRefinedInSwift = mutableListOf<CallableMemberDescriptor>()
+        for (overriddenDescriptor in descriptor.overriddenDescriptors) {
+            val (superIsRefinedForObjC, superIsRefinedInSwift) = overriddenDescriptor.inheritsRefinedAnnotations()
+            if (superIsRefinedForObjC) isRefinedForObjC = true else supersNotRefinedForObjC.add(overriddenDescriptor)
+            if (superIsRefinedInSwift) isRefinedInSwift = true else supersNotRefinedInSwift.add(overriddenDescriptor)
+        }
+        if (isRefinedForObjC && supersNotRefinedForObjC.isNotEmpty()) {
+            context.trace.reportIncompatibleOverride(declaration, objCAnnotations, supersNotRefinedForObjC)
+        }
+        if (isRefinedInSwift && supersNotRefinedInSwift.isNotEmpty()) {
+            context.trace.reportIncompatibleOverride(declaration, swiftAnnotations, supersNotRefinedInSwift)
+        }
+    }
+
+    private fun CallableMemberDescriptor.inheritsRefinedAnnotations(): Pair<Boolean, Boolean> {
```

Comment:
Why do we need to check overrides recursively? Is there a case that won't be covered by simply looking for mismatch between a method and its direct super method?

Reply:
Yes in [this](https://github.com/JetBrains/kotlin/pull/4818/files#diff-0c432a0f5175969cadff3ccfb46ea94201b7f707b5004a23aed4377fe8dcd515R108) case the direct super method doesn't have the annotations, but the two overridden methods from the interaces do.

Reply:
Right!